### PR TITLE
add  Pure :: (* -> * -> *) -> (* -> *) type family alongside the stan…

### PR DESCRIPTION
**Pull request by cartazio**
…dard Mutable type family.

This is to allow better bidirectional inference in code that mixes Mutable and Pure Vectors.

also the first of a few different changes / additions we agreed on doing for vector 0.11

theres a few few style questions that should probably be discussed before this gets merged in, such as my choice in using a CPP approach as shorthand in modules that already were doing CPP hacks.

@dolio  @tibbe  @erikd  et al would love some feedback please :)

### CODE REVIEW

**Comment by cartazio** [comment]
@ekmett i know this is a PR you've wanted for a long time :)

**Comment by cartazio** [comment]
hrm, my ` (v ~ Pure (Mutable v), MVector (Mutable v) a)` contraint on the Vector class would mean we'd have to drop ghc 7.0  and 7.2 support, but that constraint is needed to give (newer) ghc enough info to actually type check stuff.

**Comment by cartazio** [comment]
if this PR gets accepted, we'll need to raise the lowerbounds on base / ghc

**Comment by cartazio** [comment]
the ghc head build is giving some interesting warning / feedback 
eg 

```Data/Vector/Fusion/Bundle.hs:114:1: warning:
    Tab character found here.
    Please use spaces instead.
```

**Comment by cartazio** [comment]
that said, i'm not sure if the notion of redundant constraints is correct, eg 
```
Data/Vector/Generic.hs:708:11: warning:
    Redundant constraint: v ~ Pure mv
    In the type signature for:
       create :: (Vector v a, v ~ Pure mv, mv ~ Mutable v) =>
                 (forall s. ST s (mv s a)) -> v a
```

if the `v ~ Pure mv` constraint were dropped, then given only `mv`, we'd not be able to deduce `v`! Or am I overlooking something

File: Data/Vector/Fusion/Bundle/Monadic.hs [hunk 1]
```diff
@@ -107,7 +107,7 @@ import Data.Word ( Word8, Word16, Word32, Word, Word64 )
 #include "vector.h"
 #include "MachDeps.h"
 
-data Chunk v a = Chunk Int (forall m. (PrimMonad m, Vector v a) => Mutable v (PrimState m) a -> m ())
+data Chunk v a = Chunk Int (forall m mv. (PrimMonad m, Vector v a, MUTABLE_PURE(mv,v)) => mv (PrimState m) a -> m ())
```
Comment (cartazio) [inline Data/Vector/Fusion/Bundle/Monadic.hs]: i'm not sure if this change makes any difference what soever, but it seemed like a good move from a consistency perspective

**Comment by cartazio** [comment]
it looks like adding `class (v ~ Mutable (Pure v)) => MVector v a where` to `MVector` will force all the pairs of MVector/Vector instances to be in shared internal modules

**Comment by cartazio** [comment]
hrm, i dont understand the choice of pieces of code where ghc head trips over redundant constraint warnings. i'll dig into that more when i'm done travelling this weekend

**Comment by cartazio** [comment]
@dolio we never did add this did we? lets revisit this for 0.12 and i'll clean it up

File: Data/Vector/Generic/Base.hs [hunk 2]
```diff
@@ -45,19 +50,19 @@ type family Mutable (v :: * -> *) :: * -> * -> *
...
 --
-class MVector (Mutable v) a => Vector v a where
+class (v ~ Pure (Mutable v),MVector (Mutable v) a) => Vector v a where
   -- | /Assumed complexity: O(1)/
   --
   -- Unsafely convert a mutable vector to its immutable version
   -- without copying. The mutable vector may not be used after
   -- this operation.
-  basicUnsafeFreeze :: PrimMonad m => Mutable v (PrimState m) a -> m (v a)
+  basicUnsafeFreeze :: (PrimMonad m,mv~Mutable v,v~Pure mv) => mv (PrimState m) a -> m (v a)
```
Comment (OlivierSohn) [inline Data/Vector/Generic/Base.hs]: Could `MUTABLE_PURE ` be used here?

**Comment by lehins** [comment]
If I understand correctly this PR was trying to emulate injectivity that was implemented in #160 
Closing. Feel free to reopen if you think I am misunderstood this PR

**Pull request closed without merging**